### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by arpcorrea**
This is an update on the code. It creates all the field objects necessary to run a thermal analysis (conduction-convection). Although it is currently not creating the necessary fields for other models (like intimate contact development and/or healing), it is currently straightforward to create them. 
One relevant information is on how "mesh" identifies "top or bottom parts". By comparing y0, y1 and Total_thickness, it is possible to determine where to assign the boundary conditions.

### CODE REVIEW

File: consolidate/mesher/fields.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        
+         VarConv = np.concatenate((var, np.zeros((np.shape(var)[0],2))), axis=1)
+         nx_var = np.shape(VarConv)[1]
+         VarConv= np.concatenate((VarConv, np.zeros((2, nx_var))), axis=0)
+        
+
+         nx=0
+         nyu=0
+         nyl=0
+         if self.name == "Heat Input":
```
Comment (ilyasst) [inline consolidate/mesher/fields.py:26]: I do not think we need to hard code these values here.

File: consolidate/mesher/fields.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+             for BC in problem.BoundaryConditions:
+                 if BC.kind == "Heating":
```
Comment (ilyasst) [inline consolidate/mesher/fields.py:28]: We should not need to have an if for each possible condition I believe.

File: consolidate/mesher/fields.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                     for i, domains in enumerate(problem.domains):
+                         nyaux=(domains.Number_of_Elements_in_Y)+nyaux
+                         if domains.name == BC.domain:
+                             nx = domains.Number_of_Elements_in_X
+                             nyl = nyaux-i
+                             nyu = nyaux+(domains.Number_of_Elements_in_Y-i)
+                         var[nyl:nyu,0:nx] = BC.value
+                         
+                         
+         elif self.name == "Initial Convection Temperature":
```
Comment (ilyasst) [inline consolidate/mesher/fields.py:39]: I believe that all of this part can be written in a simpler way (this is not the OOP approach).

File: consolidate/mesher/mesher.py [hunk 4]
```diff
@@ -1,16 +1,31 @@
...
         self.meshes = []
+        nx=0
+        ny=0
         for domain in problem.domains:
-            self.meshes.append( Mesh(deck, domain) )
-
-        
+            self.meshes.append( Mesh(domain) )
+            nx = max(nx,domain.Number_of_Elements_in_X)
+            ny = ny + domain.Number_of_Elements_in_Y
```
Comment (ilyasst) [inline consolidate/mesher/mesher.py:19]: This line is actually problematic. This summation might not be valid for all kinds of problems ...

File: consolidate/mesher/fields.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            
+class Field:
+     def __init__(self, name, shape, problem):
+        self.name = name
+        self.populate_field(shape, problem)
+
+                    
+     def populate_field(self, shape, problem):
+         # import pdb; pdb.set_trace()
+         self.var = shape.copy()
```
[the remark this answers is not in the document]
Comment (arpcorrea) [reply consolidate/mesher/fields.py:13]: > The current approach to apply the boundary conditions is too complicated and does not really use the advantages of classes. I would like to suggest a simpler way to apply boundary conditions and domains.
> This is a Minimal Working Example that illustrates the idea:
> 
> ```python
> import numpy as np
> 
> class RectangularDomain:
> 
>     def __init__(self, name, corner0, corner1, initial_temperature ):
>         self.x0 = float(corner0[0])
>         self.y0 = float(corner0[1])
>         self.x1 = float(corner1[0])
>         self.y1 = float(corner1[1])
>         self.Lx = self.x1 - self.x0
>         self.Ly = self.y1 - self.y0
>         self.name = name
>         self.initial_temperature=initial_temperature
>         self.initial_fields = {"Temperature": float(initial_temperature)}
> 
>     def test(self, point):
>         if point[0] >= self.x0 and point[0] <= self.x1 and point[1] >= self.y0 and point[1] <= self.y1:
>             return True
>         else:
>             return False
> 
>     def generate_mask(self, problem_M):
>         self.mask = problem_M.copy()
>         for x_i, x_ in enumerate(problem_M):
>             for y_i, y_ in enumerate(problem_M):
>                 if self.test( (x_i, y_i) ):
>                     self.mask[x_i][y_i] = 1
> 
> # Simplified version of a RectangularDomain for the sake of simplicity
> domain1 = RectangularDomain("test_domain1", (0,0), (4,4), 200.)
> # This is the matrix for the problem that is generated in mesher
> M = np.zeros((10, 5))
> # This function from RectangularDomain is not called in the init, it should be called in Field()
> domain1.generate_mask(M)
> # The domain1.mask variable now contains a matrix the same shape as M, but it has zeroes everywhere except for the elements that are part of the domain
> # I used the test function to generate this matrix
> # This matrix cannot be multiplied by the initial_temeprature for example and we can directly have the Field
> # The same strategy should be adopted for Boundary Conditions
> ```

that is really powerful... Thanks

[the remark this answers is not in the document]
Comment (ilyasst) [reply consolidate/mesher/fields.py:13]: Were you able to use this in PR #8 ?

[the remark this answers is not in the document]
Comment (arpcorrea) [reply consolidate/mesher/fields.py:13]: no, didn't implement... but I have this in mind...

File: consolidate/problems/domains.py [hunk 6]
```diff
@@ -1,17 +1,48 @@
...
+        
+    def set_field_init_value(self, field_dict):
+        for key, value in field_dict.items():
+            self.initial_fields[key] = float(value)
+        
+    def generate_mask(self, problem_M):
+        self.mask = problem_M.copy()
+        for x_i in range (0,np.shape(problem_M)[0]):
+            for y_i in range (0,np.shape(problem_M)[1]):
+                # import pdb; pdb.set_trace()
```
Comment (ilyasst) [inline consolidate/problems/domains.py:43]: tiny pdn

File: consolidate/mesher/fields.py (hunk 1 above)
Comment (arpcorrea) [reply consolidate/mesher/fields.py]: something is wrong... this is not the current version (it appears Outdated here for me). The newest version doesn`t have those hard codes.

File: consolidate/mesher/fields.py (hunk 2 above)
Comment (arpcorrea) [reply consolidate/mesher/fields.py]: something is wrong... this is not the current version (it appears Outdated here for me). The newest version doesn`t have those hard codes

File: consolidate/mesher/fields.py (hunk 3 above)
Comment (arpcorrea) [reply consolidate/mesher/fields.py]: something is wrong... this is not the current version (it appears Outdated here for me). The newest version doesn`t have those hard codes

File: consolidate/mesher/mesher.py (hunk 4 above)
Comment (arpcorrea) [reply consolidate/mesher/mesher.py]: something is wrong... this is not the current version (it appears Outdated here for me). The newest version doesn`t have those lines

Comment (ilyasst) [reply consolidate/mesher/mesher.py]: These comments were done for the last time you updated the PR.
Scroll down for the latest comments after "New changes since you last viewed".
I am now wondering if you saw this comment before 🙃

File: consolidate/problems/domains.py [hunk 7]
```diff
@@ -1,17 +1,48 @@
+import numpy as np
 class RectangularDomain:
 
-    def __init__(self, name, corner0, corner1, material ):
+    def __init__(self, name, corner0, corner1, ex0, ex1, ey0, ey1, material, initialcondition, mesh ):
```
[the remark this answers is not in the document]
Comment (arpcorrea) [reply consolidate/problems/domains.py:4]: Hi! I am afraid it is necessary to provide mesh there. The reason is that with those values I further create a mask in function of them. Would you recommend something different?

File: consolidate/mesher/mesh.py [hunk 8]
```diff
@@ -2,18 +2,33 @@
 
 class Mesh:
 
-    def __init__(self, deck, domain):
-        self.set_mesh_grid(deck, domain)
+    def __init__(self,  domain, totalNx=0, totalNy=0):
+        import pdb; pdb.set_trace
```
[the remark this answers is not in the document]
Comment (arpcorrea) [reply consolidate/mesher/mesh.py:6]: resolved

File: consolidate/problems/domains.py (hunk 6 above)
Comment (arpcorrea) [reply consolidate/problems/domains.py:43]: resolved

File: TwoPlates.yaml [hunk 9]
```diff
@@ -1,49 +1,82 @@
...
+            Left Edge:
+                Temperature: 1
+            Bottom Edge:
+                Temperature: 1
+            Right Edge:
+                Temperature: 1 
+            Top Edge:
+                Horizontal asperity ratio: 1
+                Vertical asperity ratio: 1
+        Initial Condition:
```
[the remark this answers is not in the document]
Comment (arpcorrea) [reply TwoPlates.yaml:53]: Actually I believe that we just have "Temperature" as initial condition. Input Power Density is actually a Boundary Condition.

File: consolidate/mesher/mesh.py [hunk 10]
```diff
@@ -2,18 +2,33 @@
...
+        X, Y = np.meshgrid(np.arange(domain.x0, domain.x1, self.dx), np.arange(domain.y0, domain.y1, + self.dy))
+        X=X[0,:].copy()
+        Y=Y[:,0].copy()
         self.X = X
         self.Y = Y
-        self.ny1= int(self.ny/2)
\ No newline at end of file
+        
+    def set_create_mask(self, domain):
+        domain.generate_mask(self.M.copy())
```
[the remark this answers is not in the document]
Comment (arpcorrea) [reply consolidate/mesher/mesh.py:28]: there is no self.M in problem currently, only in Mesh.

**Review by ilyasst — APPROVED** [review]
It looks good :+1:
